### PR TITLE
Add the ability to remove OOO events by clicking on a button

### PR DESCRIPTION
**Pull request by tgolen**
- [x] Held on https://github.com/Expensify/Web-Expensify/pull/36603

### Fixed Issues
$ https://github.com/Expensify/Expensify/issues/257281

### Tests
1. Sign into NewDot as an Expensify employee
2. Go to your chat with Chronos
3. Add three OOO events sometime in the future on different days:

- `ooo april 1`
- `ooo april 3`
- `ooo april 5`

4. Tell chronos `ooo list`
5. Verify that all three events are listed and they each have a Remove button
6. Click the Remove button for one of the events
7. Verify that the message is updated and the event is removed

- [x] Verify that no errors appear in the JS console

### Offline tests
1. Use the same steps as above, but be offline when you click the Remove button for an event
2. Verify that the message is greyed out (with the event removed)
3. Go online
4. Verify that the message is no longer greyed out and the message still shows the event is removed

### QA Steps
Same as the tests above

- [x] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message: https://expensify.slack.com/archives/C21FRDWCV/p1677615695731209
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is correct English and approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>

</details>

<details>
<summary>Mobile Web - Chrome</summary>

</details>

<details>
<summary>Mobile Web - Safari</summary>

</details>

<details>
<summary>Desktop</summary>

</details>

<details>
<summary>iOS</summary>

</details>

<details>
<summary>Android</summary>

</details>

### CODE REVIEW

**Comment by s77rt** [comment]
@tgolen Can you add me to have Chronos access?

**Comment by tgolen** [comment]
@s77rt Unfortunately, I don't think I can. The code only works for people on the `expensify.com` domain since it's all linked to an organization calendar in Google.

**Comment by s77rt** [comment]
@tgolen No problem, so in that case a C+ reviewer is not needed here?

**Comment by tgolen** [comment]
Ah yes, because you would not be able to test it. You can still review the code if you'd like. Also, I still need to make time to fill out the checklist and add screenshots. Since it's on HOLD for a little bit, I've bought myself a little bit of time to get that completed.

File: src/components/ReportActionItem/ChronosOOOListActions.js
Comment (s77rt) [inline src/components/ReportActionItem/ChronosOOOListActions.js:38]: Any reason `errors` are not passed here?

File: src/components/ReportActionItem/ChronosOOOListActions.js
Comment (s77rt) [inline src/components/ReportActionItem/ChronosOOOListActions.js:43]: Is it guaranteed that `event` will have `start` and `end` as objects? If not this may raise an error.

File: src/components/ReportActionItem/ChronosOOOListActions.js
Comment (s77rt) [inline src/components/ReportActionItem/ChronosOOOListActions.js:64]: This is also missing translation

File: src/components/ReportActionItem/ChronosOOOListActions.js
Comment (s77rt) [inline src/components/ReportActionItem/ChronosOOOListActions.js:66]: I'm not sure about the process here, but should not we disable the button if we have a `pendingAction`?

File: src/libs/actions/Chronos.js
Comment (s77rt) [inline src/libs/actions/Chronos.js:13]: `reportAction` is not required here, use `reportActionID` instead.

Also, it may be better to change the order of params to: `reportID, reportActionID, eventID, events`

File: src/libs/actions/Chronos.js
Comment (s77rt) [inline src/libs/actions/Chronos.js:36]: replace `...reportAction` with `originalMessage: {events}`

File: src/libs/actions/Chronos.js
Comment (s77rt) [inline src/libs/actions/Chronos.js:53]: For consistency and readability move `successData` to be before `failureData`.

**Review by s77rt — CHANGES_REQUESTED** [review]

File: src/components/ReportActionItem/ChronosOOOListActions.js
Comment (tgolen) [reply src/components/ReportActionItem/ChronosOOOListActions.js:31]: Kind of lazy, but also I figured if it's only in use internally, it didn't need translated. That's probably not a great reason and thanks for calling it out! I can get it translated.

File: src/components/ReportActionItem/ChronosOOOListActions.js
Comment (tgolen) [reply src/components/ReportActionItem/ChronosOOOListActions.js:38]: I think mostly because there won't ever be any errors from the web API.

File: src/components/ReportActionItem/ChronosOOOListActions.js
Comment (tgolen) [reply src/components/ReportActionItem/ChronosOOOListActions.js:43]: It is mostly gauranteed, yeah, but I don't mind being extra careful here and can add in some lodashGet.

File: src/components/ReportActionItem/ChronosOOOListActions.js
Comment (tgolen) [reply src/components/ReportActionItem/ChronosOOOListActions.js:66]: Hm, I don't think it is necessary to. I kind of want to test that out, but I think the idea is that you should be able to delete multiple events while being offline, and that would be OK to leave the button enabled then.

File: src/libs/actions/Chronos.js
Comment (tgolen) [reply src/libs/actions/Chronos.js:13]: You're right, good idea.

**Review by tgolen — COMMENTED** [review]
Thanks for the review!

File: src/components/ReportActionItem/ChronosOOOListActions.js
Comment (s77rt) [reply src/components/ReportActionItem/ChronosOOOListActions.js:66]: The button is per event, why would you leave the button enabled if you pressed it already?

**Comment by tgolen** [comment]
All right, I got screenshots and found some bugs with the mobile layout, so I've fixed that. I've completed the checklist too

File: src/components/ReportActionItem/ChronosOOOListActions.js
Comment (tgolen) [reply src/components/ReportActionItem/ChronosOOOListActions.js:61]: Oh, I see what you mean. The specific event is removed in the optimistic data, so the row won't be rendered at all (thus, no button to disable).

**Comment by s77rt** [comment]
@tgolen Thanks for the quick update. Please mark the request changes that have been resolved as resolved. I will check again tomorrow.

File: src/libs/actions/Chronos.js
Comment (s77rt) [inline src/libs/actions/Chronos.js:9]: Missing `*` at the start for correct JSDoc syntax.

File: src/libs/actions/Chronos.js
Comment (s77rt) [inline src/libs/actions/Chronos.js:21]: I think the use of `_.reject` may be a better fit here.
Extra: Do you think we can stop the iteration once an item has been rejected? We don't really need to keep iterating over the rest.

**Review by s77rt — CHANGES_REQUESTED** [review]

File: src/libs/actions/Chronos.js
Comment (tgolen) [reply src/libs/actions/Chronos.js:21]: Good call, I like reject() or omit()

**Comment by tgolen** [comment]
Updated! I think I got everything resolved. 

bump @PauloGasparSv

**Comment by PauloGasparSv** [comment]
Hey @tgolen was about to ask if I should wait to test!

Just tested on Web before your latest commit and things look ok up until we remove one of the events. I tried removing the "april 2" event but it failed with a `403`. I pulled from `tgolen-ooo-removebutton` in Web-E before testing.
I will pull the new changes from here, re-test and debug if it happens again.

image

**Comment by PauloGasparSv** [comment]
## Reviewer Checklist

- [ ] I have verified the author checklist is complete (all boxes are checked off).
- [ ] I verified the correct issue is linked in the `### Fixed Issues` section above
- [ ] I verified testing steps are clear and they cover the changes made in this PR
    - [ ] I verified the steps for local testing are in the `Tests` section
    - [ ] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [ ] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [ ] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [ ] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I verified tests pass on **all platforms** & I tested again on:
    - [ ] Android / native
    - [ ] Android / Chrome
    - [ ] iOS / native
    - [ ] iOS / Safari
    - [ ] MacOS / Chrome / Safari
    - [ ] MacOS / Desktop
- [ ] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [ ] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [ ] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [ ] I verified that comments were added to code that is not self explanatory
    - [ ] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [ ] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [ ] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [ ] I verified any copy / text that was added to the app is correct English and approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [ ] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [ ] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [ ] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [ ] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [ ] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [ ] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [ ] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [ ] If a new component is created I verified that:
    - [ ] A similar component doesn't exist in the codebase
    - [ ] All props are defined accurately and each prop has a `/** comment above it */`
    - [ ] The file is named correctly
    - [ ] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [ ] The only data being stored in the state is data necessary for rendering and nothing else
    - [ ] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [ ] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [ ] All JSX used for rendering exists in the render method
    - [ ] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [ ] If any new file was added I verified that:
    - [ ] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [ ] If a new CSS style is added I verified that:
    - [ ] A similar style doesn't already exist
    - [ ] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [ ] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [ ] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [ ] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [ ] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [ ] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>



</details>

<details>
<summary>Mobile Web - Chrome</summary>



</details>

<details>
<summary>Mobile Web - Safari</summary>



</details>

<details>
<summary>Desktop</summary>



</details>

<details>
<summary>iOS</summary>



</details>

<details>
<summary>Android</summary>



</details>

**Review by s77rt — DISMISSED** [review]
LGTM! :rocket:

**Comment by tgolen** [comment]
@PauloGasparSv You'll also need to be on the right Auth branch as well. I think that's where that error is coming from. The auth branch is in this PR: https://github.com/Expensify/Auth/pull/7448

**Comment by PauloGasparSv** [comment]
Thks @tgolen! Re-testing here then :D

**Review by PauloGasparSv — DISMISSED** [review]
LGTM!

**Comment by tgolen** [comment]
Great! Wanna go ahead and merge @PauloGasparSv ? I only added Yuwen as a courtesy, I don't think his review is required.

**Comment by tgolen** [comment]
oh wait, it's on HOLD still. ha, nevermind

**Comment by yuwenmemon** [comment]
But the Auth branch is not deployed yet, no?

**Comment by tgolen** [comment]
Resolved conflicts. The Auth PR is on production, the Web-E PR is on staging. Once the Web-E PR is on production, I'll remove the hold here and this can be merged.

**Comment by tgolen** [comment]
@yuwenmemon This is off hold and can be merged now if you'd like to do the honors

**Review by yuwenmemon — APPROVED** [review]

**Comment by Julesssss** [comment]
Performed internalQA on each platform to unblock the deployer checklist. All tested well 👍

**Pull request merged**
